client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,338 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
+ */
+
+use smbioslib;
+use time::macros::format_description;
+use time::Date;
+
+use crate::constants::PROC_DEVICE_TREE_DIRPATH;
+use crate::models::devices;
+
+pub fn load_smbios_data(
+    entry_filepath: Option<&'static str>,
+    table_filepath: Option<&'static str>,
+) -> Result<Option<smbioslib::SMBiosData>, Box<dyn std::error::Error>> {
+    let entry_file = entry_filepath.unwrap_or(smbioslib::SYS_ENTRY_FILE);
+    let table_file = table_filepath.unwrap_or(smbioslib::SYS_TABLE_FILE);
+
+    let smbios_data = match table_load_from_device(entry_file, table_file) {
+        Ok(data) => Some(data),
+        Err(e) => {
+            eprintln!("Failed to load SMBIOS data: {}.", e);
+            None
+        }
+    };
+    Ok(smbios_data)
+}
+
+pub fn collect_bios_info(
+    smbios_data: &smbioslib::SMBiosData,
+) -> Result<devices::Bios, Box<dyn std::error::Error>> {
+    let bios_info = &smbios_data.collect::<smbioslib::SMBiosInformation>()[0];
+
+    let release_date_str = bios_info.release_date().to_string();
+    let release_date_format = format_description!("[month]/[day]/[year]");
+    let release_date_parsed = Date::parse(&release_date_str, &release_date_format)?;
+    let output_format = format_description!("[year]-[month]-[day]");
+    let release_date = release_date_parsed.format(&output_format)?;
+
+    let bios = devices::Bios {
+        firmware_revision: match (
+            bios_info.system_bios_major_release(),
+            bios_info.system_bios_minor_release(),
+        ) {
+            (Some(major), Some(minor)) => Some(format!("{}.{}", major, minor)),
+            _ => None,
+        },
+        release_date: Some(release_date),
+        revision: Some(bios_info.version().to_string()),
+        vendor: bios_info.vendor().to_string(),
+        version: bios_info.version().to_string(),
+    };
+    Ok(bios)
+}
+
+/// Retrieve CPU information from SMBios
+pub fn collect_processor_info_smbios(
+    smbios_data: &smbioslib::SMBiosData,
+    max_cpu_frequency_filepath: Option<&'static str>,
+) -> Result<devices::Processor, Box<dyn std::error::Error>> {
+    let processor_info = &smbios_data.collect::<smbioslib::SMBiosProcessorInformation>()[0];
+    let cpu_id = super::cpuid::get_cpuid(processor_info)?;
```

Comment:
Explicit imports at the top of the file instead of local `super` makes it easier to see where things are being imported and used.

Reply:
Fixed in fed34d